src/nimblepkg/publish.nim

```diff
@@ -0,0 +1,191 @@
+        extraHeaders=("Authorization: Basic $1\c\L" % a.token) &
+          "Content-Type: application/x-www-form-urlencoded\c\L" &
+          "Accept: */*\c\L")
+    let j = parseJson(x)
+    result = searchFork(j)
+  except JsonParsingError, IOError:
+    result = false
+
+proc createFork(a: Auth) =
+  discard postContent("https://api.github.com/repos/nim-lang/packages/forks",
+      extraHeaders=("Authorization: Basic $1\c\L" % a.token) &
+        "Content-Type: application/x-www-form-urlencoded\c\L" &
+        "Accept: */*\c\L")
+
+proc createPullRequest(a: Auth; packageName: string) =
+  echo "creating PR"
+  discard postContent("https://api.github.com/repos/nim-lang/packages/pulls",
+      extraHeaders=("Authorization: Basic $1\c\L" % a.token) &
+        "Content-Type: application/x-www-form-urlencoded\c\L" &
+        "Accept: */*\c\L",
+      body="""{"title": "Add package $#", "head": "$#:master",
+               "base": "master"}""" % [packageName, a.user])
+
+proc `%`(s: openArray[string]): JsonNode =
+  result = newJArray()
+  for x in s: result.add(%x)
+
+proc cleanupWhitespace(s: string): string =
+  ## Removes trailing whitespace and normalizes line endings to LF.
+  result = newStringOfCap(s.len)
+  var i = 0
+  while i < s.len:
+    if s[i] == ' ':
+      var j = i+1
+      while s[j] == ' ': inc j
+      if s[j] == '\c':
+        inc j
+        if s[j] == '\L': inc j
+        result.add '\L'
+        i = j
+      elif s[j] == '\L':
+        result.add '\L'
+        i = j+1
+      else:
+        result.add ' '
+        inc i
+    elif s[i] == '\c':
+      inc i
+      if s[i] == '\L': inc i
+      result.add '\L'
+    elif s[i] == '\L':
+      result.add '\L'
+      inc i
+    else:
+      result.add s[i]
+      inc i
+
+proc editJson(p: PackageInfo; url, tags, downloadMethod: string) =
+  var contents = parseFile("packages.json")
+  doAssert contents.kind == JArray
+  contents.add(%{
+    "name": %p.name,
+    "url": %url,
+    "method": %downloadMethod,
+    "tags": %tags.split(),
+    "description": %p.description,
+    "license": %p.license,
+    "web": %url})
+  writeFile("packages.json", contents.pretty.cleanupWhitespace)
+
+proc getPackageOriginUrl(a: Auth): string =
+  ## Adds 'user:pw' to the URL so that the user is not asked *again* for it.
+  ## We need this for 'git push'.
+  let (output, exitCode) = doCmdEx("git config --get remote.origin.url")
+  result = "origin"
+  if exitCode == 0:
+    result = output.string.strip
+    if result.endsWith(".git"): result.setLen(result.len - 4)
+    if result.startsWith("https://"):
+      result = "https://" & a.user & ':' & a.pw & '@' &
+          result["https://".len .. ^1]
+
+proc publish*(p: PackageInfo) =
+  ## Publishes the package p.
+  let auth = getGithubAuth()
+  let parent = os.getCurrentDir().parentDir()
+  var pkgsDir = parent / "nimble-packages-fork"
+  if not forkExists(auth):
+    createFork(auth)
+    echo "waiting 10s to let Github create a fork ..."
```

Comment:
True but works in practice and github recommends 5 seconds. But on a second though, we can busy wait for 1 minute, checking from time to time 'forkExists'. But I think it's good enough for now. Let people report problems.